### ISSUE TITLE
Support switching between extensions/v1beta1 and networking.k8s.io/v1 Ingress

### ISSUE THREAD

**Issue by alexmt** [argoproj/argo-rollouts#1507]
# Summary

Rollout uses `extensions/v1beta1` `Ingress` that is now longer available in v1.22. However the replacement `networking.k8s.io/v1` is available only since v1.19 . 

https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

It is too early to drop support for <v1.19 clusters so we need to support both.

# Details

Introduce controller CLI flag `--extensions-ingress` (better name is very welcome). If flag is set to true then controller should use `extensions/v1beta1` to manage Ingress. Otherwise, controller should use `networking.k8s.io/v1`

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by jessesuen** [comment]
> better name is very welcome

For consistency with our other resource kinds where we need to switch API versions (e.g. `--istio-api-version`, `--ambassador-api-version`), we should probably name this flag:

```
--ingress-api-version networking.k8s.io/v1
```

**Comment by leoluz** [comment]
Im investigating the code on how to implement this change. The reason why `istio` and `ambassador` expose a parameter to configure the api version is because they both leverage the k8s dynamic client. Using the dynamic client the API version string is the only requirement in order to correctly setup the `GVR` to interact with the resource. In the case of the Ingress however the interaction is with the resource is made is the static k8s client which means that the api version as a parameter is not going to be very useful. It seems that in order to implement this issue what we need is to update the k8s go client version used by the rollouts to 1.12. 

@jessesuen @alexmt Can you confirm is my understanding is correct? What would be the implications of updating the k8s client in the project?

**Comment by jessesuen** [comment]
> we need is to update the k8s go client version used by the rollouts to 1.12. 

We already use `k8s.io/client-go v0.20.4` (aka K8s v1.20.4) which should have both the `extensions` Ingress client and `networking.k8s.io`, so I don't think it's a matter of upgrading the go client.

> Im not sure how the newer client handles the Ingress in the old api-group. Maybe it is backwards compatible and no extra-configuration is needed

In past experience (e.g. Deployment apps vs. Deployment extensions), using the clientset for newer group (`networking.k8s.io`) would still work to modify the legacy group/kind, so I believe it should be possible to use `networking.k8s.io` with no consequence as long as the cluster supports it.

**Comment by leoluz** [comment]
> so I don't think it's a matter of upgrading the go client

You are right. client-go v0.20.4 provides both versions indeed.

> so I believe it should be possible to use networking.k8s.io with no consequence as long as the cluster supports it.

I created [this small project](https://github.com/leoluz/ingress-test) to validate this and indeed it is possible to use the client from the new `networking.k8s.io` group to manipulate the old one in `extensions/v1beta1`. Unfortunately as @alexmt pointed out, it is probably too early to drop to support of Rollouts for clusters < 1.19.

Im trying to implement this feature in the most isolated way to make it simpler to drop the `extensions/v1beta1` support in the future.

**Comment by jessesuen** [comment]
Yes, I suspect you may need to implement an abstraction over the two ingress clientsets which basically does:

```golang
if ingressGroup == "extensions" {
    c.client.ExtensionsV1beta1().Ingresses(ingress.Namespace).Update(ctx, newIngress, metav1.UpdateOptions{})
} else {
    c.client.NetworkingV1().Ingresses(ingress.Namespace).Update(ctx, newIngress, metav1.UpdateOptions{})
}
```

**Comment by leoluz** [comment]
This is the first draft implementing an Ingress Wrapper in the `ingress` package. @jessesuen @alexmt if you agree with this direction I will update the PR replacing all `Ingress` usage in the project to use this pattern.
https://github.com/argoproj/argo-rollouts/pull/1529

**Comment by jinstest1** [comment]
when we are expecting the release of the above bugfix ?

**Comment by kvendingoldo** [comment]
can it be backported to 1.1.x?

**Comment by kvendingoldo** [comment]
any news?

**Comment by golgotow** [comment]
how is possible to install argo-rollouts to 1.22 with no issue to ingress v1beta1 ?

**Comment by kvendingoldo** [comment]
can it be backported to 1.1.x version? can we reopen this issue, pls?

**Comment by leoluz** [comment]
Argo-Rollouts 1.2 will be released soon. 
If this can't wait, please consider using the `latest` image.
